tests/sccache_cargo.rs

```diff
@@ -3,53 +3,101 @@
 //! Any copyright is dedicated to the Public Domain.
 //! http://creativecommons.org/publicdomain/zero/1.0/
 
-#![deny(rust_2018_idioms)]
+use anyhow::{Context, Result};
+use once_cell::sync::Lazy;
+
+use assert_cmd::prelude::*;
+use chrono::Local;
+use predicates::prelude::*;
+use std::ffi::OsString;
+use std::fs;
+use std::io::Write;
+use std::path::{Path, PathBuf};
+use std::process::{Command, Stdio};
 
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
 #[macro_use]
 extern crate log;
 
-/// Test that building a simple Rust crate with cargo using sccache results in a cache hit
-/// when built a second time.
-#[test]
-#[cfg(all(not(target_os = "windows"), not(target_os = "macos")))]
-fn test_rust_cargo() {
-    test_rust_cargo_cmd("check", &[]);
-    test_rust_cargo_cmd("build", &[]);
+static SCCACHE_BIN: Lazy<PathBuf> = Lazy::new(|| assert_cmd::cargo::cargo_bin("sccache"));
+static CARGO: Lazy<OsString> = Lazy::new(|| std::env::var_os("CARGO").unwrap());
+static CRATE_DIR: Lazy<PathBuf> =
+    Lazy::new(|| Path::new(file!()).parent().unwrap().join("test-crate"));
 
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("check", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
-    #[cfg(feature = "unstable")]
-    test_rust_cargo_cmd("build", &[("RUSTFLAGS", std::ffi::OsStr::new("-Zprofile"))]);
+/// Used as a test setup fixture. The drop implementation cleans up after a _successfull_ test.
+/// It is therefore important that the asserts use `Result` to return errors instead of panicking.
+/// Otherwise the temporary directory will not be cleaned up and the started sccache server process
+/// will not be stopped.
+struct SccacheTest<'a> {
+    /// Tempdir used for Sccache cache and cargo output. It is kept in the struct only to have the
+    /// destructor run when SccacheTest goes out of scope, but is never used otherwise.
+    #[allow(dead_code)]
+    tempdir: tempfile::TempDir,
```

Comment:
Perhaps have a `run_cargo()` method on `SccacheTest` that builds the `env` dict so that the compiler can more easily understand the lifecycle of the `tempdir`. As it is now, we still have "references" to the `tempdir` in the form of raw string paths, but this is too abstract for the compiler to track the lifetime automatically.

Reply:
I added the `show_stats()` impl on the SccacheTest struct, so that the struct is used when running `sccache --show-stats`. 

I double checked, and [the reference](https://doc.rust-lang.org/reference/destructors.html#scopes-of-function-parameters) guarantees that function parameters are dropped last, at the very end of the function. Struct members also can't be dropped prematurely, so I would prefer to not complicate things and just keep it as is.